Unconditional return in obfs4 Transport

go linter Identified an unconditional return in the obfs4 transport in a loop that should check all obfs4 registrations associated with a specific phantom IP address.  

```
application/transports/wrapping/obfs4/obfs4.go:64:3: SA4004: the surrounding loop is unconditionally terminated (staticcheck)
                return r, wrapped, err
```
There should likely be a `continue` condition somewhere in this for loop 

https://github.com/refraction-networking/conjure/blob/40a148614703f0ebe2e9c5b3522bb7bb11e0600f/application/transports/wrapping/obfs4/obfs4.go#L33-L65

--- a/application/lib/test/phantom_subnets_min.toml
+++ b/application/lib/test/phantom_subnets_min.toml
@@ -1,0 +1,12 @@
+
+# USED BY:
+# * lib/phantom_selector_test.go
+# * transports/wrapping/obfs4/obfs4_test.go
+#
+# Do not change without consulting those files and ensuring tests still pass.
+[Networks]
+	[Networks.1]
+        Generation = 1
+        [[Networks.1.WeightedSubnets]]
+            Weight = 1
+            Subnets = ["192.122.190.0/32", "2001:48a8:687f:1::/128"]

--- a/application/transports/wrapping/internal/tests/tests.go
+++ b/application/transports/wrapping/internal/tests/tests.go
@@ -18,8 +18,14 @@ type Transport struct {
 
 var SharedSecret = []byte(`6a328b8ec2024dd92dd64332164cc0425ddbde40cb7b81e055bf7b099096d068`)
 
+// SetupPhantomConnections registers one session with the provided transport and
+// registration manager using a pre-determined kay and phantom subnet file.
 func SetupPhantomConnections(manager *dd.RegistrationManager, transport pb.TransportType) (clientToPhantom net.Conn, serverFromPhantom net.Conn, reg *dd.DecoyRegistration) {
 	testSubnetPath := os.Getenv("GOPATH") + "/src/github.com/refraction-networking/conjure/application/lib/test/phantom_subnets.toml"
+	return SetupPhantomConnectionsSecret(manager, transport, SharedSecret, testSubnetPath)
+}
+
+func SetupPhantomConnectionsSecret(manager *dd.RegistrationManager, transport pb.TransportType, sharedSecret []byte, testSubnetPath string) (clientToPhantom net.Conn, serverFromPhantom net.Conn, reg *dd.DecoyRegistration) {
 	os.Setenv("PHANTOM_SUBNET_LOCATION", testSubnetPath)
 
 	phantom := bufconn.Listen(65535)
@@ -44,7 +50,7 @@ func SetupPhantomConnections(manager *dd.RegistrationManager, transport pb.Trans
 
 	wg.Wait()
 
-	keys, err := dd.GenSharedKeys(SharedSecret)
+	keys, err := dd.GenSharedKeys(sharedSecret)
 	if err != nil {
 		log.Fatalln("failed to generate shared keys:", err)
 	}

--- a/application/transports/wrapping/obfs4/obfs4.go
+++ b/application/transports/wrapping/obfs4/obfs4.go
@@ -3,6 +3,7 @@ package obfs4
 import (
 	"bytes"
 	"fmt"
+	"io/ioutil"
 	"net"
 
 	pt "git.torproject.org/pluggable-transports/goptlib.git"
@@ -30,18 +31,11 @@ func (Transport) WrapConnection(data *bytes.Buffer, c net.Conn, phantom net.IP, 
 	var representative ntor.Representative
 	copy(representative[:ntor.RepresentativeLength], data.Bytes()[:ntor.RepresentativeLength])
 
-	// TODO: This seems to be an issue since we return unconditionally so why is it a loop?
-	// should something be checked so we continue to more registrations?
 	for _, r := range getObfs4Registrations(regManager, phantom) {
 		mark := generateMark(r.Keys.Obfs4Keys.NodeID, r.Keys.Obfs4Keys.PublicKey, &representative)
 		pos := findMarkMac(mark, data.Bytes(), ntor.RepresentativeLength+ClientMinPadLength, MaxHandshakeLength, true)
-
 		if pos == -1 {
-			// If we read up to the max handshake length and didn't find the mark, move on.
-			if data.Len() >= MaxHandshakeLength {
-				return nil, nil, transports.ErrNotTransport
-			}
-			return nil, nil, transports.ErrTryAgain
+			continue
 		}
 
 		// We found the mark in the client handshake! We found our registration!
@@ -55,7 +49,12 @@ func (Transport) WrapConnection(data *bytes.Buffer, c net.Conn, phantom net.IP, 
 		args.Add("drbg-seed", seed.Hex())
 
 		t := &obfs4.Transport{}
-		factory, err := t.ServerFactory("", &args)
+		stateDir, err := ioutil.TempDir("", "")
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create tmp-dir for WrapConn")
+		}
+
+		factory, err := t.ServerFactory(stateDir, &args)
 		if err != nil {
 			return nil, nil, fmt.Errorf("failed to create server factory: %w", err)
 		}
@@ -64,6 +63,13 @@ func (Transport) WrapConnection(data *bytes.Buffer, c net.Conn, phantom net.IP, 
 		wrapped, err := factory.WrapConn(mc)
 
 		return r, wrapped, err
+	}
+
+	// If we read more than min handshake len, but less than max and didn't find
+	// the mark get more bytes until we have reached the max handshake length.
+	// If we have reached the max handshake len and didn't find it return NotTransport
+	if data.Len() < MaxHandshakeLength {
+		return nil, nil, transports.ErrTryAgain
 	}
 
 	// The only time we'll make it here is if there are no obfs4 registrations

--- a/application/transports/wrapping/obfs4/obfs4_test.go
+++ b/application/transports/wrapping/obfs4/obfs4_test.go
@@ -11,6 +11,7 @@ import (
 	"testing"
 	"time"
 
+	dd "github.com/refraction-networking/conjure/application/lib"
 	"github.com/refraction-networking/conjure/application/transports"
 	"github.com/refraction-networking/conjure/application/transports/wrapping/internal/tests"
 	pb "github.com/refraction-networking/gotapdance/protobuf"
@@ -20,14 +21,14 @@ import (
 	"gitlab.com/yawning/obfs4.git/transports/obfs4"
 )
 
-func wrapConnection(conn net.Conn, nodeID, publicKey string, wrapped chan (net.Conn)) {
+func wrapConnection(conn net.Conn, nodeID, publicKey string, wrapped chan (net.Conn), stateDir string) {
 	args := pt.Args{}
 	args.Add("node-id", nodeID)
 	args.Add("public-key", publicKey)
 	args.Add("iat-mode", "1")
 
 	t := obfs4.Transport{}
-	c, err := t.ClientFactory("")
+	c, err := t.ClientFactory(stateDir)
 	if err != nil {
 		log.Fatalln("failed to set up client factory:", err)
 	}
@@ -57,7 +58,83 @@ func TestSuccessfulWrap(t *testing.T) {
 	defer sfp.Close()
 
 	wrappedc2p := make(chan net.Conn)
-	go wrapConnection(c2p, reg.Keys.Obfs4Keys.NodeID.Hex(), reg.Keys.Obfs4Keys.PublicKey.Hex(), wrappedc2p)
+	stateDir := t.TempDir()
+	go wrapConnection(c2p, reg.Keys.Obfs4Keys.NodeID.Hex(), reg.Keys.Obfs4Keys.PublicKey.Hex(), wrappedc2p, stateDir)
+
+	var buf [4096]byte
+	var buffer bytes.Buffer
+	var wrappedsfp net.Conn
+	var err error
+	for {
+		n, _ := sfp.Read(buf[:])
+		buffer.Write(buf[:n])
+
+		_, wrappedsfp, err = transport.WrapConnection(&buffer, sfp, reg.DarkDecoy, manager)
+		if errors.Is(err, transports.ErrTryAgain) {
+			continue
+		} else if err != nil {
+			t.Fatalf("expected nil or ErrTryAgain, got %v", err)
+		}
+
+		break
+	}
+
+	select {
+	case c2p = <-wrappedc2p:
+	case <-time.After(100 * time.Millisecond):
+		t.Fatal("timed out waiting for wrapped client connection")
+	}
+
+	message := []byte(`test message!`)
+	_, err = c2p.Write(message)
+	require.Nil(t, err)
+
+	received := make([]byte, len(message))
+	_, err = io.ReadFull(wrappedsfp, received)
+	if err != nil {
+		t.Fatalf("failed reading from connection: %v", err)
+	}
+
+	if !bytes.Equal(message, received) {
+		t.Fatalf("expected %v, got %v", message, received)
+	}
+}
+
+// TestSuccessfulWrapMulti is designed to ensure that the obfs4 implementation
+// of WrapConnection works when multiple registrations might share the same
+// phantom IP address. This requires mapping multiple registrations to the same
+// phantom before calling WrapConnection on one of them. This is done using a
+// subnet file that only has a an /32 allocation meaning all registrations by
+// default map to the same phantom address (for n=5 registrations). We wrap the
+// last connection.
+func TestSuccessfulWrapMulti(t *testing.T) {
+	testSubnetPath := os.Getenv("GOPATH") + "/src/github.com/refraction-networking/conjure/application/lib/test/phantom_subnets_min.toml"
+	os.Setenv("PHANTOM_SUBNET_LOCATION", testSubnetPath)
+
+	sharedSecrets := [][]byte{
+		[]byte(`b07c17169ac6c4ec77de4b795e939e3994dc708be1afb6bcd5e646941cf97f35`),
+		[]byte(`3143952b9355b6187ddc6104eb9ea85fca52ba5f8d88a93f73910f860b133217`),
+		[]byte(`62a807d9f89673960564185e893530216ef889545d9c039f9b02d8ccd36c093d`),
+		[]byte(`d078a5084786cfd51094e6c27718451e1260285068b6dbc7009c06de38d49711`),
+		tests.SharedSecret,
+	}
+
+	var transport Transport
+	manager := tests.SetupRegistrationManager(tests.Transport{Index: pb.TransportType_Obfs4, Transport: transport})
+	var c2p, sfp net.Conn
+	var reg *dd.DecoyRegistration
+
+	// register 5 sessions guaranteeing collisions on phantom IP addresses
+	for _, secret := range sharedSecrets {
+		c2p, sfp, reg = tests.SetupPhantomConnectionsSecret(manager, pb.TransportType_Obfs4, secret, testSubnetPath)
+	}
+
+	defer c2p.Close()
+	defer sfp.Close()
+
+	wrappedc2p := make(chan net.Conn)
+	stateDir := t.TempDir()
+	go wrapConnection(c2p, reg.Keys.Obfs4Keys.NodeID.Hex(), reg.Keys.Obfs4Keys.PublicKey.Hex(), wrappedc2p, stateDir)
 
 	var buf [4096]byte
 	var buffer bytes.Buffer

